src/api/block_builder/input_selection/utxo_chains.rs

```diff
@@ -0,0 +1,218 @@
+                                {
+                                    let address =
+                                        Address::try_from(&immutable_alias_address_unlock_condition_dto.address)?;
+                                    utxo_chains.push((address, output_response.clone()));
+                                }
+                            }
+                        }
+                    }
+                }
+                _ => {}
+            }
+        }
+
+        // Get recursively owned alias or nft outputs
+        get_alias_and_nft_outputs_recursively(self.client, &mut utxo_chains).await?;
+
+        let mut utxo_chain_inputs = Vec::new();
+        for (unlock_address, output_response) in utxo_chains {
+            let (address_index, internal) = match self.secret_manager {
+                Some(secret_manager) => {
+                    match unlock_address {
+                        Address::Ed25519(_) => {
+                            search_address(
+                                secret_manager,
+                                &bech32_hrp,
+                                self.coin_type,
+                                self.account_index,
+                                self.input_range.clone(),
+                                &unlock_address,
+                            )
+                            .await?
+                        }
+                        // Alias and NFT addresses can't be generated from a private key
+                        _ => (0, false),
+                    }
+                }
+                None => (0, false),
+            };
+
+            utxo_chain_inputs.push(InputSigningData {
+                output: Output::try_from(&output_response.output)?,
+                output_metadata: OutputMetadata::try_from(&output_response.metadata)?,
+                chain: Some(Chain::from_u32_hardened(vec![
+                    HD_WALLET_TYPE,
+                    self.coin_type,
+                    self.account_index,
+                    internal as u32,
+                    address_index,
+                ])),
+                bech32_address: unlock_address.to_bech32(&bech32_hrp),
+            });
+        }
+
+        Ok(utxo_chain_inputs)
+    }
+}
+
+/// Get recursively owned alias and nft outputs and add them to the utxo_chains
+pub(crate) async fn get_alias_and_nft_outputs_recursively(
+    client: &Client,
+    utxo_chains: &mut Vec<(Address, OutputResponse)>,
+) -> Result<()> {
+    log::debug!("[get_alias_and_nft_outputs_recursively]");
+    let current_time = client.get_time_checked().await?;
+
+    let mut unprocessed_alias_nft_addresses = std::collections::HashSet::new();
+
+    for (unlock_address, _output_response) in utxo_chains.clone() {
+        unprocessed_alias_nft_addresses.insert(unlock_address);
+    }
+
+    while !unprocessed_alias_nft_addresses.is_empty() {
+        for (unlock_address, _output_response) in utxo_chains.clone() {
```

Comment:
I feel like we can massively improve this by having a second vector instead of a Hashset. We pop-iterate on `utxo_chains` and add processed things to the second vector, at the end we set the first vector to the second. No more N^2.